File: Algorithms/0114_Flatten_Binary_Tree_to_Linked_List/Python/Flatten_Binary_Tree_to_Linked_List_Solution_1.py

# Space: O(n)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def flatten(self, root):
        """
        Do not return anything, modify root in-place instead.
        """
        if root is None: return
        if root.left is None and root.right is None: return
        stack = [root]

        while stack:
            cur = stack.pop()
            if cur.right: stack.append(cur.right)
            if cur.left: stack.append(cur.left)
            if stack: cur.right = stack[-1]
            cur.left = None

File: Algorithms/0114_Flatten_Binary_Tree_to_Linked_List/Java/FlattenBinaryTreetoLinkedListSolution1.java

// Space: O(n)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        if (root.left == null && root.right == null) return;
        Stack<TreeNode> stack = new Stack<>();
        stack.add(root);

        while (stack.size() > 0) {
            TreeNode cur = stack.pop();
            if (cur.right != null) stack.add(cur.right);
            if (cur.left != null) stack.add(cur.left);
            if (stack.size() > 0) cur.right = stack.peek();
            cur.left = null;
        }
    }
}




